SignalServiceKit/src/Messages/OWSReadReceiptManager.m

```diff
@@ -445,6 +412,148 @@ - (void)updateOutgoingMessageFromLinkedDevice:(TSOutgoingMessage *)message
+            OWSLinkedDeviceReadReceipt *readReceipt = [[OWSLinkedDeviceReadReceipt alloc] initWithSenderId:senderId timestamp:timestamp];
+            [readReceipt saveWithTransaction:transaction];
+        }
+    }
+}
+
+- (void)markAsReadOnLinkedDevice:(TSIncomingMessage *)message
+                     transaction:(YapDatabaseReadWriteTransaction *)transaction
+{
+    OWSAssert(message);
+    OWSAssert(transaction);
+    
+    [self markAsReadBeforeTimestamp:message.timestamp
+                             thread:message.thread
+                           wasLocal:NO
+                        transaction:transaction];
+}
+
+#pragma mark - Mark As Read
+
+- (void)markAsReadBeforeTimestamp:(uint64_t)timestamp
+                           thread:(TSThread *)thread
+                         wasLocal:(BOOL)wasLocal
+                      transaction:(YapDatabaseReadWriteTransaction *)transaction
+{
+    OWSAssert(timestamp > 0);
+    OWSAssert(thread);
+    OWSAssert(transaction);
+    
+    NSMutableArray<id<OWSReadTracking>> *interactions = [NSMutableArray new];
+    
+    [[TSDatabaseView unseenDatabaseViewExtension:transaction]
+     enumerateRowsInGroup:thread.uniqueId
+     usingBlock:^(NSString *collection,
+                  NSString *key,
+                  id object,
+                  id metadata,
+                  NSUInteger index,
+                  BOOL *stop) {
+         
+         if (![object conformsToProtocol:@protocol(OWSReadTracking)]) {
+             OWSFail(
+                     @"Expected to conform to OWSReadTracking: object with class: %@ collection: %@ "
+                     @"key: %@",
+                     [object class],
+                     collection,
+                     key);
+             return;
+         }
+         id<OWSReadTracking> possiblyRead = (id<OWSReadTracking>)object;
+         
+         if (possiblyRead.timestampForSorting > timestamp) {
+             *stop = YES;
+             return;
+         }
+         
+         OWSAssert(!possiblyRead.read);
+         if (!possiblyRead.read) {
+             [interactions addObject:possiblyRead];
+         }
+     }];
+    
+    if (interactions.count < 1) {
+        return;
+    }
+    
+    if (wasLocal) {
+        DDLogError(@"Marking %zd messages as read locally.", interactions.count);
+    } else {
+        DDLogError(@"Marking %zd messages as read by linked device.", interactions.count);
+    }
+    for (id<OWSReadTracking> possiblyRead in interactions) {
+        [possiblyRead markAsReadWithTransaction:transaction sendReadReceipt:wasLocal updateExpiration:YES];
+        
+        if ([possiblyRead isKindOfClass:[TSIncomingMessage class]]) {
+            TSIncomingMessage *incomingMessage = (TSIncomingMessage *)possiblyRead;
+            dispatch_async(dispatch_get_main_queue(), ^{
+                [[NSNotificationCenter defaultCenter]
+                 postNotificationNameAsync:kMessageMarkedAsReadNotification
```

Comment:
nit:  Since we now have a concept of marking **outgoing** messages as read, but this notification only applies to **incoming** messages, maybe we should reflect that in the name.

`kMessageMarkedAsReadNotification` -> `kIncomingMessageMarkedAsReadNotification`

Reply:
Great idea.